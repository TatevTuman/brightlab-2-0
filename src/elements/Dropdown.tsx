import React, { Fragment, memo, useRef } from 'react'
import { Menu, Transition } from '@headlessui/react'
import { Children, OptionType } from '~types'
import { useOutsideClick } from '~hooks'

type DropdownProps = {
  show: boolean
  children: Children
  options?: OptionType[]
  groups?: { title?: string; options: OptionType[] }[]
  onShow: () => void
  onHide: () => void
}

const Dropdown: React.FC<DropdownProps> = props => {
  const { show, options, groups, children, onShow, onHide } = props

  const isOptions = options && options.length
  const isGroupedOptions = groups && groups.length

  const renderMenuGroupedOptions = () => {
    if (!isGroupedOptions) return null

    return (
      <Menu.Items
        static
        className="w-full max-h-200 overflow-y-scroll absolute left-0 mt-2 origin-top-right bg-white divide-y divide-gray-100 rounded-6 shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none"
      >
        {groups!.map((group, index) => {
          const { title, options } = group

          return (
            <div key={index} className="px-1 py-1">
              {title && <h6>{title}</h6>}
              {options.map(option => {
                const { label, value } = option

                return (
                  <Menu.Item key={value}>
                    {({ active }) => (
                      <button
                        className={`${
                          active ? 'bg-violet-500 text-white' : 'text-gray-900'
                        } group flex rounded-md items-center w-full px-2 py-2 text-14`}
                      >
                        {label}
                      </button>
                    )}
                  </Menu.Item>
                )
              })}
            </div>
          )
        })}
      </Menu.Items>
    )
  }

  const renderMenuOptions = () => {
    if (!isOptions) return null

    return (
      <Menu.Items
        static
        className="w-full max-h-200 overflow-y-scroll absolute left-0 mt-2 origin-top-right bg-white rounded-6 shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none"
      >
        {options!.map(option => {
          const { label, value } = option

          return (
            <Menu.Item key={value}>
              {({ active }) => (
                <button
                  className={`${
                    active ? 'bg-violet-500 text-white' : 'text-gray-900'
                  } group flex rounded-md items-center w-full px-2 py-2 text-14`}
                >
                  {label}
                </button>
              )}
            </Menu.Item>
          )
        })}
      </Menu.Items>
    )
  }

  const render = () => {
    if (isGroupedOptions) return renderMenuGroupedOptions()
    else if (isOptions) return renderMenuOptions()
    else return 'No options supplied'
  }

  const ref = useRef<HTMLDivElement>(null)
  useOutsideClick(ref, onHide)

  return (
    <div className="relative inline-block text-left" ref={ref}>
      <Menu as={Fragment}>
        {({ open }) => (
          <>
            <div className="inline-flex justify-center cursor-pointer" onClick={onShow}>
              {children}
            </div>
            <Transition
              as={'div'}
              show={show}
              enter="transition ease-out duration-200"
              enterFrom="transform opacity-0 scale-90"
              enterTo="transform opacity-100 scale-100"
              leave="transition ease-in duration-150"
              leaveFrom="transform opacity-100 scale-100"
              leaveTo="transform opacity-0 scale-90"
            >
              {render()}
            </Transition>
          </>
        )}
      </Menu>
    </div>
  )
}

export default memo(Dropdown)
