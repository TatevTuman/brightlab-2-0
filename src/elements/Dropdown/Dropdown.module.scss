.dropdown {
  list-style: none;
  margin: 0.5rem 0 0.5rem calc(1rem + 3px);
}
