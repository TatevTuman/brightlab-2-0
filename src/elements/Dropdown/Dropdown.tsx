import React, { memo } from 'react'
import { OptionType } from '@types'
import styles from './Dropdown.module.scss'

export interface DropdownProps<T> {
  id: string
  options: OptionType<T>[]
  onSelect: (option: OptionType<T>) => void
}

const Dropdown = <T,>(props: DropdownProps<T>) => {
  const { id, options, onSelect } = props

  const MemoizedDropdown = memo(() => {
    return (
      <ul id={id} className={styles.dropdown} tabIndex={0}>
        {options.map((option, index) => {
          const { label } = option
          const key = label + index

          return (
            <li onClick={() => onSelect(option)} key={key} tabIndex={0}>
              {label}
            </li>
          )
        })}
      </ul>
    )
  })

  MemoizedDropdown.displayName = 'Dropdown'

  return <MemoizedDropdown />
}

export default Dropdown
