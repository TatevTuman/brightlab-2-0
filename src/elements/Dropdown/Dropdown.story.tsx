import React from 'react'
import { Meta, Story } from '@storybook/react/types-6-0'
import { action } from '@storybook/addon-actions'
import Dropdown, { DropdownProps } from './Dropdown'

interface DropdownStoryTemplateProps extends Omit<DropdownProps, 'options'> {
  optionText: string
}

export default {
  title: 'Elements/Dropdown',
  argTypes: {
    toggle: { control: 'boolean', name: 'Toggle' },
    optionText: { control: 'text', name: 'Option Text' },
    onSelect: { action: 'onSelect' }
  }
} as Meta

const DropdownStoryTemplate: Story<DropdownStoryTemplateProps> = args => {
  const options = new Array(100).fill({ label: args.optionText, value: args.optionText })
  return <Dropdown {...args} options={options} />
}

DropdownStoryTemplate.args = {
  optionText: 'Option',
  toggle: true,
  onSelect: action('onSelect')
}

export const DropdownStory = DropdownStoryTemplate.bind({})
DropdownStory.args = {
  ...DropdownStoryTemplate.args
}
