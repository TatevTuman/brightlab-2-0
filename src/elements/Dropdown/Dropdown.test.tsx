import React from 'react'
import { cleanup, render, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Dropdown from './Dropdown'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Dropdown', () => {
  const options = [{ value: 'test', label: 'Test' }]

  it('renders correctly', async () => {
    // const { container } = render(<Dropdown options={options} onSelect={() => null} />)
    // const awaitedContainer = await waitFor(() => container)
    //
    // expect(awaitedContainer).toMatchSnapshot()
    expect(0).toBe(1)
  })
})
