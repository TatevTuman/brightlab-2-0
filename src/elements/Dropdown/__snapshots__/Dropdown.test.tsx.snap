// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders correctly 1`] = `
<div>
  <ul
    class="dropdown"
  >
    <li>
      Test
    </li>
  </ul>
</div>
`;
